src/main/tetris_board.cpp

```diff
@@ -106,28 +113,59 @@ void TetrisBoard::add_piece( TetrisPiece& piece )
   int piece_current_row = piece.get_current_row();
   int piece_current_column = piece.get_current_column();
 
-  for( auto& point : piece.get_block_locations() )
+  vector<unique_ptr<Point>>& block_locations = piece.get_block_locations();
+  vector<unique_ptr<RenderComponent>>& render_components = piece.get_render_components();
+  for( size_t i = 0; i < block_locations.size(); i++ )
   {
+    auto& point = block_locations.at( i );
+    
     int occupied_x = point -> get_x() + piece_current_column;
     int occupied_y = point -> get_y() + piece_current_row;
 
     occupied_spaces
       .at( occupied_y )
       .at( occupied_x ) = true;
-  }
 
-  for( auto& component : piece.get_render_components() )
-  {
-    components.push_back( move( component ) );
+    try
+    {
+      Block& block = dynamic_cast<Block&>( *render_components.at( i ) );
```

Comment:
The TetrisBoard stores the Blocks as pointers to RenderComponents, so to manipulate the components as blocks I have to downcast. This seems crude, but I haven't thought of a more elegant means to handle this issue yet.